#87 add video widget

PhilippMDoerner opened the pull request:
A draft PR for #87 .

I *think* this is mostly finished.
It provides a Video widget with setters so you can provide a fileName (string) or a file (GFile) to play in it.
It... I don't think it's free of memory leaks, I see the memory consumption of the example application grow as I play around with the filename, but sometimes it also drastically jumps down again ? Honestly not sure if that means we're free of leaks or not.

I free the old media stream if a new media stream gets set via property hook at least.

This is draft because I don't yet have a strategy on how to free `GtkMediaStream` memory when the Video widget gets destroyed.
Until that is cleared up I do not consider this mergeable.
A possible strategy is outlined in my last comment on #87 but I'd like a sanity check for what I *think* could work there.

can-lehmann commented:
For freeing the `MediaStream` you want to do something like `Pixbuf`/`TextBuffer`: https://github.com/can-lehmann/owlkettle/blob/main/owlkettle/widgets.nim#L464

PhilippMDoerner commented:
Hmm, so define a destructor got GtkMediaStream and when Video gets destroyed and garbage collected it inevitably also triggers the GtkMediaStream destructor, causing a call to g_object_unref which free's the `GtkMediaStream` ?
I'll try it.

Should be relatively easy though, basically just defining a destructor and a container type

PhilippMDoerner commented:
> For freeing the `MediaStream` you want to do something like `Pixbuf`/`TextBuffer`: https://github.com/can-lehmann/owlkettle/blob/main/owlkettle/widgets.nim#L464

Okay, did that. I've:
- Wrapped `GtkMediaStream` in `MediaStreamObj` whose ref-type alias is `MediaStream`
- Defined destructors and copy-hook for `MediaStreamObj`
- Created a bunch of constructors for `MediaStream` (GFile and Filename can be used)
- Changed the example so that it provides a filechooser dialog which fetches filenames instead of having a string autoform field
- Added a reset-button so you can test if the destructor gets called

The mechanism appears to work as intended. There still seems to be very much a memory leak, but I'm not quite sure where it's coming from. At least in my testing, flipping between 2 screen cast files easily balloons the memory consumption from 60mb to > 130mb

can-lehmann commented on examples/widgets/video.nim:
```diff
@@ -0,0 +1,92 @@
+                app.mediaStream = nil
+        
+      Box(orient = OrientY):
+        Label(text = "Video via FileName") {.expand: false.}
```
I would remove this.

can-lehmann commented on examples/widgets/video.nim:
```diff
@@ -0,0 +1,92 @@
+              except IoError as err:
+                echo err.msg
+                app.mediaStream = nil
+        
```
Add simple `play`, `pause`, `skip forward`, `skip backward` media controls on the right side of the HeaderBar. These should demonstrate the available functions on `MediaStream`.

PhilippMDoerner replied:
Done

PhilippMDoerner commented on owlkettle/widgets.nim:
```diff
@@ -3525,6 +3525,98 @@ renderable Scale of BaseWidget:
+# See TODO at comment of PixbufObj regarding why we wrap GtkMediaStream with MediaStreamObj
+type 
+  MediaStreamObj = object
+    gdk: GtkMediaStream
```
In that case quick question regarding pattern:
In the other objects we called this field gdk. 
Is the fieldname supposed to resemble the name of the library where the object is from?

can-lehmann replied:
The idea is basically: `MediaStream` is the "Owlkettle object" and `MediaStream.gtk` is the "GTK object" belonging to that owlkettle object. It could also be called something like `native`, `internal` (as with internalWidget) or `extern`, but the naming convention for the other wrapper objects was using `gdk`/`gtk` depending on the types origin.

PhilippMDoerner commented on examples/widgets/video.nim:
```diff
@@ -0,0 +1,92 @@
+      title = "Video Example"
+      HeaderBar() {.addTitlebar.}:
+        insert(app.toAutoFormMenu(ignoreFields = @["mediaStream"], sizeRequest = (400, 400))) {.addRight.}
+        
```
Done

PhilippMDoerner commented on owlkettle/widgets.nim:
```diff
@@ -3525,6 +3525,98 @@ renderable Scale of BaseWidget:
+      echo "Obj: ", isNil(state.mediaStream)
+      echo "Widget: ", isNil(state.internalWidget)
+      if not isNil(state.mediaStream):
+        echo "Stream: ", isNil(state.mediaStream.gdk)
```
This was me still figuring out where possible memory leaks might be coming from.
I still can observe growing memory useage as the source of the video gets changed over and over, but I can't figure out what supposedly is leaking and leading to increased RAM useage.

can-lehmann reviewed (commented):
Overall, really good work on this one!

Some final changes I made while playing with the example:
- `gint64` maps to `int64` in Nim, not `cint` as `cint` may be of a different size e.g. on 32bit systems
- Getters and setters for e.g. `timestamp` should be named `timestamp` and `timestamp=` as to allow for easy use in Nim
- `ButtonFlat` should be used in header bars
